README.md
# nDPI FFI binding for LuaJIT

`ljndpi` is a Lua [FFI](http://luajit.org/ext_ffi.html) binding for the
[nDPI][ndpi] deep packet inspection library.

## Usage

The following program skeleton outlines the usage of the module:

```lua
local ndpi = require "ndpi"
local TICK_RESOLUTION = 1000  -- Samples per second
local detector = ndpi.detection_module(TICK_RESOLUTION)
detector:set_protocol_bitmask(ndpi.protocol_bitmask():set_all())

local flows = {}
for packet in iterate_packets() do
  -- Obtain a unique identifier for a flow, most likely using
  -- the source/destination IP addresses, ports, and VLAN tag.
  local flow_id = get_flow_id(packet)
  local flow = flows[flow_id]
  if not flow then
    -- Add any other flow-specific information needed by your application.
    flow = {
      ndpi_flow = ndpi.flow(),
      ndpi_src_id = ndpi.id(),
      ndpi_dst_id = ndpi.id(),
      protocol = ndpi.protocol.PROTOCOL_UNKNOWN,
    }
    flows[flow_id] = flow
  end

  if flow.protocol ~= ndpi.protocol.PROTOCOL_UNKNOWN then
    print("Identified protocol: " .. ndpi.protocol[flow.protocol])
  else
    flow.protocol = detector:process_packet(flow.ndpi_flow,
                                            get_ip_data_pointer(packet),
                                            get_ip_data_length(packet),
                                            os.time(),
                                            flow.ndpi_src_id,
                                            flow.ndpi_dst_id)
  end
end
```

## Requirements

* [LuaJIT 2.0](http://www.luajit.org) or later.
* [nDPI 1.7][ndpi] or later, versions 1.7, 1.8, 2.0, 2.2, 2.3, 2.4, and 3.4
  are known to work. For other versions you may need to use the
  [update-protocool-ids](tools/update-protocol-ids) script to generate
  a `protocol_ids_*.lua` file for it.

## Installation

Using [LuaRocks](https://luarocks.org) is recommended. The latest stable
release can be installed with:

```sh
luarocks install ljndpi
```

The current development version can be installed with:

```sh
luarocks install --server=https://luarocks.org/dev ljndpi
```

Alternatively, you can just place the `ndpi` subdirectory and the `ndpi.lua`
file in any location where LuaJIT will be able to find them.


## Documentation

None yet. The API follows that of the nDPI C library loosely, building on
metatypes to provide a more idiomatic feeling. For the moment the best option
is to check the [example
program](https://github.com/aperezdc/ljndpi/blob/master/examples/readpcap).

The following table summarizes the equivalence between C and Lua types:

| Lua Type | C Type |
|:---------|:-------|
| `ndpi.detection_module` | `struct ndpi_detection_module_struct` |
| `ndpi.flow` | `struct ndpi_flow_struct` |
| `ndpi.id` | `struct ndpi_id_struct` |
| `ndpi.protocol_bitmask` | `NDPI_PROTOCOL_BITMASK` |

As for the functions, they can be accessed Lua's method invocation syntax
(`foo:bar()`), for example the following C code:

```c
#define TICKS 1000
NDPI_PROTOCOL_BITMASK all_bits;
NDPI_BITMASK_SET_ALL(all_bits);
struct ndpi_detection_module_struct *dm =
        ndpi_init_detection_module(TICKS, malloc, free, NULL);
ndpi_set_protocol_detection_bitmask2(dm, &all_bits);
```

becomes:

```lua
local TICKS = 1000
local all_bits = ndpi.protocol_bitmask()
all_bits:set_all()
local dm = ndpi.detection_module(TICKS)
dm:set_protocol_bitmask(all_bits)
```

Note that many methods return the objects themselves, allowing for chained
method calls, which allows the above snippet to be simplified into:

```lua
local TICKS = 1000
local dm = ndpi.detection_module(TICKS):set_protocol_bitmask(ndpi.protocol_bitmask():set_all())
```

The “nDPI QuickStart Guide” from the
[ntop documentation downloads
section](http://www.ntop.org/support/documentation/documentation/) is helpful
to get overview of how to use nDPI, which also applies to using `ljndpi`.


## Bugs

Check the [issue tracker](https://github.com/aperezdc/ljndpi/issues) for known
bugs, and please file a bug if you find one.


## License

All the `ljndpi` code is under the [Apache
License 2.0](http://www.apache.org/licenses/). A copy of the license is
included in the [COPYING](COPYING) file in the source distribution.

### Historical Note

Versions prior to `0.1.0` are under the terms of the [MIT/X11
license](https://opensource.org/licenses/mit). Versions `0.1.0` and `0.0.3`
only differ in their license.


## Authors

`ljndpi` was written by Adrián Pérez at [Igalia S.L.](http://www.igalia.com),
initially as part of [SnabbWall](http://snabbwall.org). Development was
sponsored by the [NLnet Foundation](https://nlnet.nl/).

![](http://snabbwall.org/images/igalia-logo.png) &nbsp;&nbsp;
![](http://snabbwall.org/images/nlnet-logo.gif)

Feedback is very welcome! If you are interested in `ljndpi` in a Snabb
context, probably the best thing is to post a message to the
[snabb-devel](https://groups.google.com/forum/#!forum/snabb-devel) group. Or,
if you like, you can contact Adrián directly at `aperez@igalia.com`. If you
have a problem that `ljndpi` can help solve, let us know!

[ndpi]: http://www.ntop.org/products/deep-packet-inspection/ndpi/


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.2.0] - 2020-11-05
### Added
- Add support for nDPI versions 2.0, 2.2, 2.3, and 3.4

## [v0.1.0] - 2017-03-21
### Changed
- License changed from MIT/X11 to the Apache License 2.0

## [v0.0.3] - 2016-07-23
### Added
* Support using version 1.8 of `libndpi`.
* Values of type `ndpi.protocol_bitmask` now have a `__tostring` metamethod.
* The module is now installable using [LuaRocks](https://luarocks.org).

## [v0.0.2] - 2016-02-29
### Fixed
* Avoid calls to `ffi.gc()` passing `nil` as second argument for flows and
  identifiers.

## v0.0.1 - 2016-01-24
* First release.

[Unreleased]: https://github.com/aperezdc/ljndpi/compare/v0.2.0...HEAD
[v0.2.0]: https://github.com/aperezdc/ljndpi/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/aperezdc/ljndpi/compare/v0.0.3...v0.1.0
[v0.0.3]: https://github.com/aperezdc/ljndpi/compare/v0.0.2...v0.0.3
[v0.0.2]: https://github.com/aperezdc/ljndpi/compare/v0.0.1...v0.0.2


ndpi/protocol_ids_3_4.lua
-- Generated by ljdnpi's tools/update-protocol-ids script
local T = {
  [0] = "PROTOCOL_UNKNOWN",
  [1] = "PROTOCOL_FTP_CONTROL",
  [2] = "PROTOCOL_MAIL_POP",
  [3] = "PROTOCOL_MAIL_SMTP",
  [4] = "PROTOCOL_MAIL_IMAP",
  [5] = "PROTOCOL_DNS",
  [6] = "PROTOCOL_IPP",
  [7] = "PROTOCOL_HTTP",
  [8] = "PROTOCOL_MDNS",
  [9] = "PROTOCOL_NTP",
  [10] = "PROTOCOL_NETBIOS",
  [11] = "PROTOCOL_NFS",
  [12] = "PROTOCOL_SSDP",
  [13] = "PROTOCOL_BGP",
  [14] = "PROTOCOL_SNMP",
  [15] = "PROTOCOL_XDMCP",
  [16] = "PROTOCOL_SMBV1",
  [17] = "PROTOCOL_SYSLOG",
  [18] = "PROTOCOL_DHCP",
  [19] = "PROTOCOL_POSTGRES",
  [20] = "PROTOCOL_MYSQL",
  [21] = "PROTOCOL_HOTMAIL",
  [22] = "PROTOCOL_DIRECT_DOWNLOAD_LINK",
  [23] = "PROTOCOL_MAIL_POPS",
  [24] = "PROTOCOL_APPLEJUICE",
  [25] = "PROTOCOL_DIRECTCONNECT",
  [26] = "PROTOCOL_NTOP",
  [27] = "PROTOCOL_COAP",
  [28] = "PROTOCOL_VMWARE",
  [29] = "PROTOCOL_MAIL_SMTPS",
  [30] = "PROTOCOL_FBZERO",
  [31] = "PROTOCOL_UBNTAC2",
  [32] = "PROTOCOL_KONTIKI",
  [33] = "PROTOCOL_OPENFT",
  [34] = "PROTOCOL_FASTTRACK",
  [35] = "PROTOCOL_GNUTELLA",
  [36] = "PROTOCOL_EDONKEY",
  [37] = "PROTOCOL_BITTORRENT",
  [38] = "PROTOCOL_SKYPE_CALL",
  [39] = "PROTOCOL_SIGNAL",
  [40] = "PROTOCOL_MEMCACHED",
  [41] = "PROTOCOL_SMBV23",
  [42] = "PROTOCOL_MINING",
  [43] = "PROTOCOL_NEST_LOG_SINK",
  [44] = "PROTOCOL_MODBUS",
  [45] = "PROTOCOL_WHATSAPP_CALL",
  [46] = "PROTOCOL_DATASAVER",
  [47] = "PROTOCOL_XBOX",
  [48] = "PROTOCOL_QQ",
  [49] = "PROTOCOL_TIKTOK",
  [50] = "PROTOCOL_RTSP",
  [51] = "PROTOCOL_MAIL_IMAPS",
  [52] = "PROTOCOL_ICECAST",
  [53] = "PROTOCOL_FREE_53",
  [54] = "PROTOCOL_PPSTREAM",
  [55] = "PROTOCOL_ZATTOO",
  [56] = "PROTOCOL_SHOUTCAST",
  [57] = "PROTOCOL_SOPCAST",
  [58] = "PROTOCOL_DISCORD",
  [59] = "PROTOCOL_TVUPLAYER",
  [60] = "PROTOCOL_HTTP_DOWNLOAD",
  [61] = "PROTOCOL_QQLIVE",
  [62] = "PROTOCOL_THUNDER",
  [63] = "PROTOCOL_SOULSEEK",
  [64] = "PROTOCOL_PS_VUE",
  [65] = "PROTOCOL_IRC",
  [66] = "PROTOCOL_AYIYA",
  [67] = "PROTOCOL_UNENCRYPTED_JABBER",
  [68] = "PROTOCOL_NATS",
  [69] = "PROTOCOL_FREE_69",
  [70] = "PROTOCOL_YAHOO",
  [71] = "PROTOCOL_FREE_71",
  [72] = "PROTOCOL_GOOGLE_PLUS",
  [73] = "PROTOCOL_IP_VRRP",
  [74] = "PROTOCOL_STEAM",
  [75] = "PROTOCOL_HALFLIFE2",
  [76] = "PROTOCOL_WORLDOFWARCRAFT",
  [77] = "PROTOCOL_TELNET",
  [78] = "PROTOCOL_STUN",
  [79] = "PROTOCOL_IP_IPSEC",
  [80] = "PROTOCOL_IP_GRE",
  [81] = "PROTOCOL_IP_ICMP",
  [82] = "PROTOCOL_IP_IGMP",
  [83] = "PROTOCOL_IP_EGP",
  [84] = "PROTOCOL_IP_SCTP",
  [85] = "PROTOCOL_IP_OSPF",
  [86] = "PROTOCOL_IP_IP_IN_IP",
  [87] = "PROTOCOL_RTP",
  [88] = "PROTOCOL_RDP",
  [89] = "PROTOCOL_VNC",
  [90] = "PROTOCOL_FREE90",
  [91] = "PROTOCOL_TLS",
  [92] = "PROTOCOL_SSH",
  [93] = "PROTOCOL_USENET",
  [94] = "PROTOCOL_MGCP",
  [95] = "PROTOCOL_IAX",
  [96] = "PROTOCOL_TFTP",
  [97] = "PROTOCOL_AFP",
  [98] = "PROTOCOL_STEALTHNET",
  [99] = "PROTOCOL_AIMINI",
  [100] = "PROTOCOL_SIP",
  [101] = "PROTOCOL_TRUPHONE",
  [102] = "PROTOCOL_IP_ICMPV6",
  [103] = "PROTOCOL_DHCPV6",
  [104] = "PROTOCOL_ARMAGETRON",
  [105] = "PROTOCOL_CROSSFIRE",
  [106] = "PROTOCOL_DOFUS",
  [107] = "PROTOCOL_FIESTA",
  [108] = "PROTOCOL_FLORENSIA",
  [109] = "PROTOCOL_GUILDWARS",
  [110] = "PROTOCOL_HTTP_ACTIVESYNC",
  [111] = "PROTOCOL_KERBEROS",
  [112] = "PROTOCOL_LDAP",
  [113] = "PROTOCOL_MAPLESTORY",
  [114] = "PROTOCOL_MSSQL_TDS",
  [115] = "PROTOCOL_PPTP",
  [116] = "PROTOCOL_WARCRAFT3",
  [117] = "PROTOCOL_WORLD_OF_KUNG_FU",
  [118] = "PROTOCOL_SLACK",
  [119] = "PROTOCOL_FACEBOOK",
  [120] = "PROTOCOL_TWITTER",
  [121] = "PROTOCOL_DROPBOX",
  [122] = "PROTOCOL_GMAIL",
  [123] = "PROTOCOL_GOOGLE_MAPS",
  [124] = "PROTOCOL_YOUTUBE",
  [125] = "PROTOCOL_SKYPE",
  [126] = "PROTOCOL_GOOGLE",
  [127] = "PROTOCOL_DCERPC",
  [128] = "PROTOCOL_NETFLOW",
  [129] = "PROTOCOL_SFLOW",
  [130] = "PROTOCOL_HTTP_CONNECT",
  [131] = "PROTOCOL_HTTP_PROXY",
  [132] = "PROTOCOL_CITRIX",
  [133] = "PROTOCOL_NETFLIX",
  [134] = "PROTOCOL_LASTFM",
  [135] = "PROTOCOL_WAZE",
  [136] = "PROTOCOL_YOUTUBE_UPLOAD",
  [137] = "PROTOCOL_HULU",
  [138] = "PROTOCOL_CHECKMK",
  [139] = "PROTOCOL_AJP",
  [140] = "PROTOCOL_APPLE",
  [141] = "PROTOCOL_WEBEX",
  [142] = "PROTOCOL_WHATSAPP",
  [143] = "PROTOCOL_APPLE_ICLOUD",
  [144] = "PROTOCOL_VIBER",
  [145] = "PROTOCOL_APPLE_ITUNES",
  [146] = "PROTOCOL_RADIUS",
  [147] = "PROTOCOL_WINDOWS_UPDATE",
  [148] = "PROTOCOL_TEAMVIEWER",
  [149] = "PROTOCOL_TUENTI",
  [150] = "PROTOCOL_LOTUS_NOTES",
  [151] = "PROTOCOL_SAP",
  [152] = "PROTOCOL_GTP",
  [153] = "PROTOCOL_UPNP",
  [154] = "PROTOCOL_LLMNR",
  [155] = "PROTOCOL_REMOTE_SCAN",
  [156] = "PROTOCOL_SPOTIFY",
  [157] = "PROTOCOL_MESSENGER",
  [158] = "PROTOCOL_H323",
  [159] = "PROTOCOL_OPENVPN",
  [160] = "PROTOCOL_NOE",
  [161] = "PROTOCOL_CISCOVPN",
  [162] = "PROTOCOL_TEAMSPEAK",
  [163] = "PROTOCOL_TOR",
  [164] = "PROTOCOL_SKINNY",
  [165] = "PROTOCOL_RTCP",
  [166] = "PROTOCOL_RSYNC",
  [167] = "PROTOCOL_ORACLE",
  [168] = "PROTOCOL_CORBA",
  [169] = "PROTOCOL_UBUNTUONE",
  [170] = "PROTOCOL_WHOIS_DAS",
  [171] = "PROTOCOL_COLLECTD",
  [172] = "PROTOCOL_SOCKS",
  [173] = "PROTOCOL_NINTENDO",
  [174] = "PROTOCOL_RTMP",
  [175] = "PROTOCOL_FTP_DATA",
  [176] = "PROTOCOL_WIKIPEDIA",
  [177] = "PROTOCOL_ZMQ",
  [178] = "PROTOCOL_AMAZON",
  [179] = "PROTOCOL_EBAY",
  [180] = "PROTOCOL_CNN",
  [181] = "PROTOCOL_MEGACO",
  [182] = "PROTOCOL_REDIS",
  [183] = "PROTOCOL_FREE_183",
  [184] = "PROTOCOL_VHUA",
  [185] = "PROTOCOL_TELEGRAM",
  [186] = "PROTOCOL_VEVO",
  [187] = "PROTOCOL_PANDORA",
  [188] = "PROTOCOL_QUIC",
  [189] = "PROTOCOL_ZOOM",
  [190] = "PROTOCOL_EAQ",
  [191] = "PROTOCOL_OOKLA",
  [192] = "PROTOCOL_AMQP",
  [193] = "PROTOCOL_KAKAOTALK",
  [194] = "PROTOCOL_KAKAOTALK_VOICE",
  [195] = "PROTOCOL_TWITCH",
  [196] = "PROTOCOL_DOH_DOT",
  [197] = "PROTOCOL_WECHAT",
  [198] = "PROTOCOL_MPEGTS",
  [199] = "PROTOCOL_SNAPCHAT",
  [200] = "PROTOCOL_SINA",
  [201] = "PROTOCOL_HANGOUT_DUO",
  [202] = "PROTOCOL_IFLIX",
  [203] = "PROTOCOL_GITHUB",
  [204] = "PROTOCOL_BJNP",
  [205] = "PROTOCOL_FREE_205",
  [206] = "PROTOCOL_WIREGUARD",
  [207] = "PROTOCOL_SMPP",
  [208] = "PROTOCOL_DNSCRYPT",
  [209] = "PROTOCOL_TINC",
  [210] = "PROTOCOL_DEEZER",
  [211] = "PROTOCOL_INSTAGRAM",
  [212] = "PROTOCOL_MICROSOFT",
  [213] = "PROTOCOL_STARCRAFT",
  [214] = "PROTOCOL_TEREDO",
  [215] = "PROTOCOL_HOTSPOT_SHIELD",
  [216] = "PROTOCOL_IMO",
  [217] = "PROTOCOL_GOOGLE_DRIVE",
  [218] = "PROTOCOL_OCS",
  [219] = "PROTOCOL_MICROSOFT_365",
  [220] = "PROTOCOL_CLOUDFLARE",
  [221] = "PROTOCOL_MS_ONE_DRIVE",
  [222] = "PROTOCOL_MQTT",
  [223] = "PROTOCOL_RX",
  [224] = "PROTOCOL_APPLESTORE",
  [225] = "PROTOCOL_OPENDNS",
  [226] = "PROTOCOL_GIT",
  [227] = "PROTOCOL_DRDA",
  [228] = "PROTOCOL_PLAYSTORE",
  [229] = "PROTOCOL_SOMEIP",
  [230] = "PROTOCOL_FIX",
  [231] = "PROTOCOL_PLAYSTATION",
  [232] = "PROTOCOL_PASTEBIN",
  [233] = "PROTOCOL_LINKEDIN",
  [234] = "PROTOCOL_SOUNDCLOUD",
  [235] = "PROTOCOL_CSGO",
  [236] = "PROTOCOL_LISP",
  [237] = "PROTOCOL_DIAMETER",
  [238] = "PROTOCOL_APPLE_PUSH",
  [239] = "PROTOCOL_GOOGLE_SERVICES",
  [240] = "PROTOCOL_AMAZON_VIDEO",
  [241] = "PROTOCOL_GOOGLE_DOCS",
  [242] = "PROTOCOL_WHATSAPP_FILES",
  [243] = "PROTOCOL_TARGUS_GETDATA",
  [244] = "PROTOCOL_DNP3",
  [245] = "PROTOCOL_IEC60870",
  [246] = "PROTOCOL_BLOOMBERG",
  [247] = "PROTOCOL_CAPWAP",
  [248] = "PROTOCOL_ZABBIX",
  [249] = "PROTOCOL_S7COMM",
  [250] = "PROTOCOL_MSTEAMS",
  [251] = "PROTOCOL_WEBSOCKET",
  [252] = "PROTOCOL_ANYDESK",
  [253] = "PROTOCOL_SOAP",
  PROTOCOL_UNKNOWN = 0,
  PROTOCOL_FTP_CONTROL = 1,
  PROTOCOL_MAIL_POP = 2,
  PROTOCOL_MAIL_SMTP = 3,
  PROTOCOL_MAIL_IMAP = 4,
  PROTOCOL_DNS = 5,
  PROTOCOL_IPP = 6,
  PROTOCOL_HTTP = 7,
  PROTOCOL_MDNS = 8,
  PROTOCOL_NTP = 9,
  PROTOCOL_NETBIOS = 10,
  PROTOCOL_NFS = 11,
  PROTOCOL_SSDP = 12,
  PROTOCOL_BGP = 13,
  PROTOCOL_SNMP = 14,
  PROTOCOL_XDMCP = 15,
  PROTOCOL_SMBV1 = 16,
  PROTOCOL_SYSLOG = 17,
  PROTOCOL_DHCP = 18,
  PROTOCOL_POSTGRES = 19,
  PROTOCOL_MYSQL = 20,
  PROTOCOL_HOTMAIL = 21,
  PROTOCOL_DIRECT_DOWNLOAD_LINK = 22,
  PROTOCOL_MAIL_POPS = 23,
  PROTOCOL_APPLEJUICE = 24,
  PROTOCOL_DIRECTCONNECT = 25,
  PROTOCOL_NTOP = 26,
  PROTOCOL_COAP = 27,
  PROTOCOL_VMWARE = 28,
  PROTOCOL_MAIL_SMTPS = 29,
  PROTOCOL_FBZERO = 30,
  PROTOCOL_UBNTAC2 = 31,
  PROTOCOL_KONTIKI = 32,
  PROTOCOL_OPENFT = 33,
  PROTOCOL_FASTTRACK = 34,
  PROTOCOL_GNUTELLA = 35,
  PROTOCOL_EDONKEY = 36,
  PROTOCOL_BITTORRENT = 37,
  PROTOCOL_SKYPE_CALL = 38,
  PROTOCOL_SIGNAL = 39,
  PROTOCOL_MEMCACHED = 40,
  PROTOCOL_SMBV23 = 41,
  PROTOCOL_MINING = 42,
  PROTOCOL_NEST_LOG_SINK = 43,
  PROTOCOL_MODBUS = 44,
  PROTOCOL_WHATSAPP_CALL = 45,
  PROTOCOL_DATASAVER = 46,
  PROTOCOL_XBOX = 47,
  PROTOCOL_QQ = 48,
  PROTOCOL_TIKTOK = 49,
  PROTOCOL_RTSP = 50,
  PROTOCOL_MAIL_IMAPS = 51,
  PROTOCOL_ICECAST = 52,
  PROTOCOL_FREE_53 = 53,
  PROTOCOL_PPSTREAM = 54,
  PROTOCOL_ZATTOO = 55,
  PROTOCOL_SHOUTCAST = 56,
  PROTOCOL_SOPCAST = 57,
  PROTOCOL_DISCORD = 58,
  PROTOCOL_TVUPLAYER = 59,
  PROTOCOL_HTTP_DOWNLOAD = 60,
  PROTOCOL_QQLIVE = 61,
  PROTOCOL_THUNDER = 62,
  PROTOCOL_SOULSEEK = 63,
  PROTOCOL_PS_VUE = 64,
  PROTOCOL_IRC = 65,
  PROTOCOL_AYIYA = 66,
  PROTOCOL_UNENCRYPTED_JABBER = 67,
  PROTOCOL_NATS = 68,
  PROTOCOL_FREE_69 = 69,
  PROTOCOL_YAHOO = 70,
  PROTOCOL_FREE_71 = 71,
  PROTOCOL_GOOGLE_PLUS = 72,
  PROTOCOL_IP_VRRP = 73,
  PROTOCOL_STEAM = 74,
  PROTOCOL_HALFLIFE2 = 75,
  PROTOCOL_WORLDOFWARCRAFT = 76,
  PROTOCOL_TELNET = 77,
  PROTOCOL_STUN = 78,
  PROTOCOL_IP_IPSEC = 79,
  PROTOCOL_IP_GRE = 80,
  PROTOCOL_IP_ICMP = 81,
  PROTOCOL_IP_IGMP = 82,
  PROTOCOL_IP_EGP = 83,
  PROTOCOL_IP_SCTP = 84,
  PROTOCOL_IP_OSPF = 85,
  PROTOCOL_IP_IP_IN_IP = 86,
  PROTOCOL_RTP = 87,
  PROTOCOL_RDP = 88,
  PROTOCOL_VNC = 89,
  PROTOCOL_FREE90 = 90,
  PROTOCOL_TLS = 91,
  PROTOCOL_SSH = 92,
  PROTOCOL_USENET = 93,
  PROTOCOL_MGCP = 94,
  PROTOCOL_IAX = 95,
  PROTOCOL_TFTP = 96,
  PROTOCOL_AFP = 97,
  PROTOCOL_STEALTHNET = 98,
  PROTOCOL_AIMINI = 99,
  PROTOCOL_SIP = 100,
  PROTOCOL_TRUPHONE = 101,
  PROTOCOL_IP_ICMPV6 = 102,
  PROTOCOL_DHCPV6 = 103,
  PROTOCOL_ARMAGETRON = 104,
  PROTOCOL_CROSSFIRE = 105,
  PROTOCOL_DOFUS = 106,
  PROTOCOL_FIESTA = 107,
  PROTOCOL_FLORENSIA = 108,
  PROTOCOL_GUILDWARS = 109,
  PROTOCOL_HTTP_ACTIVESYNC = 110,
  PROTOCOL_KERBEROS = 111,
  PROTOCOL_LDAP = 112,
  PROTOCOL_MAPLESTORY = 113,
  PROTOCOL_MSSQL_TDS = 114,
  PROTOCOL_PPTP = 115,
  PROTOCOL_WARCRAFT3 = 116,
  PROTOCOL_WORLD_OF_KUNG_FU = 117,
  PROTOCOL_SLACK = 118,
  PROTOCOL_FACEBOOK = 119,
  PROTOCOL_TWITTER = 120,
  PROTOCOL_DROPBOX = 121,
  PROTOCOL_GMAIL = 122,
  PROTOCOL_GOOGLE_MAPS = 123,
  PROTOCOL_YOUTUBE = 124,
  PROTOCOL_SKYPE = 125,
  PROTOCOL_GOOGLE = 126,
  PROTOCOL_DCERPC = 127,
  PROTOCOL_NETFLOW = 128,
  PROTOCOL_SFLOW = 129,
  PROTOCOL_HTTP_CONNECT = 130,
  PROTOCOL_HTTP_PROXY = 131,
  PROTOCOL_CITRIX = 132,
  PROTOCOL_NETFLIX = 133,
  PROTOCOL_LASTFM = 134,
  PROTOCOL_WAZE = 135,
  PROTOCOL_YOUTUBE_UPLOAD = 136,
  PROTOCOL_HULU = 137,
  PROTOCOL_CHECKMK = 138,
  PROTOCOL_AJP = 139,
  PROTOCOL_APPLE = 140,
  PROTOCOL_WEBEX = 141,
  PROTOCOL_WHATSAPP = 142,
  PROTOCOL_APPLE_ICLOUD = 143,
  PROTOCOL_VIBER = 144,
  PROTOCOL_APPLE_ITUNES = 145,
  PROTOCOL_RADIUS = 146,
  PROTOCOL_WINDOWS_UPDATE = 147,
  PROTOCOL_TEAMVIEWER = 148,
  PROTOCOL_TUENTI = 149,
  PROTOCOL_LOTUS_NOTES = 150,
  PROTOCOL_SAP = 151,
  PROTOCOL_GTP = 152,
  PROTOCOL_UPNP = 153,
  PROTOCOL_LLMNR = 154,
  PROTOCOL_REMOTE_SCAN = 155,
  PROTOCOL_SPOTIFY = 156,
  PROTOCOL_MESSENGER = 157,
  PROTOCOL_H323 = 158,
  PROTOCOL_OPENVPN = 159,
  PROTOCOL_NOE = 160,
  PROTOCOL_CISCOVPN = 161,
  PROTOCOL_TEAMSPEAK = 162,
  PROTOCOL_TOR = 163,
  PROTOCOL_SKINNY = 164,
  PROTOCOL_RTCP = 165,
  PROTOCOL_RSYNC = 166,
  PROTOCOL_ORACLE = 167,
  PROTOCOL_CORBA = 168,
  PROTOCOL_UBUNTUONE = 169,
  PROTOCOL_WHOIS_DAS = 170,
  PROTOCOL_COLLECTD = 171,
  PROTOCOL_SOCKS = 172,
  PROTOCOL_NINTENDO = 173,
  PROTOCOL_RTMP = 174,
  PROTOCOL_FTP_DATA = 175,
  PROTOCOL_WIKIPEDIA = 176,
  PROTOCOL_ZMQ = 177,
  PROTOCOL_AMAZON = 178,
  PROTOCOL_EBAY = 179,
  PROTOCOL_CNN = 180,
  PROTOCOL_MEGACO = 181,
  PROTOCOL_REDIS = 182,
  PROTOCOL_FREE_183 = 183,
  PROTOCOL_VHUA = 184,
  PROTOCOL_TELEGRAM = 185,
  PROTOCOL_VEVO = 186,
  PROTOCOL_PANDORA = 187,
  PROTOCOL_QUIC = 188,
  PROTOCOL_ZOOM = 189,
  PROTOCOL_EAQ = 190,
  PROTOCOL_OOKLA = 191,
  PROTOCOL_AMQP = 192,
  PROTOCOL_KAKAOTALK = 193,
  PROTOCOL_KAKAOTALK_VOICE = 194,
  PROTOCOL_TWITCH = 195,
  PROTOCOL_DOH_DOT = 196,
  PROTOCOL_WECHAT = 197,
  PROTOCOL_MPEGTS = 198,
  PROTOCOL_SNAPCHAT = 199,
  PROTOCOL_SINA = 200,
  PROTOCOL_HANGOUT_DUO = 201,
  PROTOCOL_IFLIX = 202,
  PROTOCOL_GITHUB = 203,
  PROTOCOL_BJNP = 204,
  PROTOCOL_FREE_205 = 205,
  PROTOCOL_WIREGUARD = 206,
  PROTOCOL_SMPP = 207,
  PROTOCOL_DNSCRYPT = 208,
  PROTOCOL_TINC = 209,
  PROTOCOL_DEEZER = 210,
  PROTOCOL_INSTAGRAM = 211,
  PROTOCOL_MICROSOFT = 212,
  PROTOCOL_STARCRAFT = 213,
  PROTOCOL_TEREDO = 214,
  PROTOCOL_HOTSPOT_SHIELD = 215,
  PROTOCOL_IMO = 216,
  PROTOCOL_GOOGLE_DRIVE = 217,
  PROTOCOL_OCS = 218,
  PROTOCOL_MICROSOFT_365 = 219,
  PROTOCOL_CLOUDFLARE = 220,
  PROTOCOL_MS_ONE_DRIVE = 221,
  PROTOCOL_MQTT = 222,
  PROTOCOL_RX = 223,
  PROTOCOL_APPLESTORE = 224,
  PROTOCOL_OPENDNS = 225,
  PROTOCOL_GIT = 226,
  PROTOCOL_DRDA = 227,
  PROTOCOL_PLAYSTORE = 228,
  PROTOCOL_SOMEIP = 229,
  PROTOCOL_FIX = 230,
  PROTOCOL_PLAYSTATION = 231,
  PROTOCOL_PASTEBIN = 232,
  PROTOCOL_LINKEDIN = 233,
  PROTOCOL_SOUNDCLOUD = 234,
  PROTOCOL_CSGO = 235,
  PROTOCOL_LISP = 236,
  PROTOCOL_DIAMETER = 237,
  PROTOCOL_APPLE_PUSH = 238,
  PROTOCOL_GOOGLE_SERVICES = 239,
  PROTOCOL_AMAZON_VIDEO = 240,
  PROTOCOL_GOOGLE_DOCS = 241,
  PROTOCOL_WHATSAPP_FILES = 242,
  PROTOCOL_TARGUS_GETDATA = 243,
  PROTOCOL_DNP3 = 244,
  PROTOCOL_IEC60870 = 245,
  PROTOCOL_BLOOMBERG = 246,
  PROTOCOL_CAPWAP = 247,
  PROTOCOL_ZABBIX = 248,
  PROTOCOL_S7COMM = 249,
  PROTOCOL_MSTEAMS = 250,
  PROTOCOL_WEBSOCKET = 251,
  PROTOCOL_ANYDESK = 252,
  PROTOCOL_SOAP = 253,
}
T.PROTOCOL_NO_MASTER_PROTO = T.PROTOCOL_UNKNOWN
return T
